### DISCUSSION TITLE
how to get `item`'s `item_location`?

### DISCUSSION THREAD

**Discussion by scarf005** [cataclysmbn/Cataclysm-BN#2572] [Q&A] [answered]
i'm planning to do #2487. the plan is:

1. player eats a food.
2. if the food has `EATEN_COLD`, check its location.
3. if its temperature is lower than certain threshold, give morale bonus akin to how `EATEN_HOT` works.

location temperature can be easily found via `rot::temperature_flag_for_location` added in #2267.
however, `Character::modify_morale` does not take `item_location` as parameter. could i get `item_location` from item? or should i add item_location parameter for all these nested function calls? ~~hecc that suspiciously resembles [prop drilling](https://blog.logrocket.com/solving-prop-drilling-react-apps/)~~ 

`Character::modify_morale` <- `Character::consume_effects` <- `Character::eat`

**Comment by joveeater** [comment, marked as answer]
You can't get an item_location from an item. You will have to pass the item_location through. You can go from an item_location TO an item though and as such rather than adding a new parameter you can replace the existing item parameter. 

This does suck, there might be a lot of nested calls need changing. You'll need to know where the item is in order to use item_location's constructors properly and that can be a lot of calls away. It's all going away as part of my refactor. item_location has been removed entirely and you can just use item everywhere.

**Reply by scarf005** [reply]
thanks! if so, i could gladly wait until #2250.

**Reply by Coolthulhu** [reply]
It's unlikely to get in soon, we'll need a LOT of testing. It will probably get its own "line" of releases before it gets fully mainlined, as massive reworks of internal logic of this kind tend to have lots of issues that need lots of testing.

For now, the correct way to do things like this is to rewrite the "call stack" to use `item_location`, AKA the hard way.

**Reply by scarf005** [reply]
alright. mmm, i'm making more spaghetti.
